go/arrow/array/data.go

```diff
@@ -63,6 +64,16 @@ func NewData(dtype arrow.DataType, length int, buffers []*memory.Buffer, childDa
 	}
 }
 
+// NewDataWithDictionary creates a new data object, but also sets the provided dictionary into the data if it's not nil
+func NewDataWithDictionary(dtype arrow.DataType, length int, buffers []*memory.Buffer, childData []arrow.ArrayData, nulls, offset int, dict *Data) *Data {
+	data := NewData(dtype, length, buffers, childData, nulls, offset)
```

Comment:
should we check that childData is empty here?

Reply:
Is it guaranteed that `childData` should always be empty for Dictionary arrays? Based on the spec, can you have a dictionary of lists where the values of the dictionary are themselves list arrays? If it's guaranteed to always be empty, I'll just remove it from the argument list here.

Reply:
also, we'd only verify that childData is empty *iff* dict is non-nil.

@emkornfield do you know the answer to whether or not `childData` should *always* be empty for a dictionary array?

Reply:
the spec allows for arbitrary dictionary values (including lists).  I don't think this is super well supported in the general ecosystem though.